clang/lib/CIR/Dialect/Transforms/TargetLowering/ItaniumCXXABI.cpp

```diff
@@ -0,0 +1,57 @@
+//===------- ItaniumCXXABI.cpp - Emit CIR code Itanium-specific code  -----===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This provides CIR lowering logic targeting the Itanium C++ ABI. The class in
+// this file generates structures that follow the Itanium C++ ABI, which is
+// documented at:
+//  https://itanium-cxx-abi.github.io/cxx-abi/abi.html
+//  https://itanium-cxx-abi.github.io/cxx-abi/abi-eh.html
+//
+// It also supports the closely-related ARM ABI, documented at:
+// https://developer.arm.com/documentation/ihi0041/g/
+//
+// This file partially mimics clang/lib/CodeGen/ItaniumCXXABI.cpp. The queries
+// are adapted to operate on the CIR dialect, however.
+//
+//===----------------------------------------------------------------------===//
+
+#include "CIRCXXABI.h"
+#include "LowerModule.h"
+
+namespace mlir {
+namespace cir {
+
+namespace {
+
+class ItaniumCXXABI : public CIRCXXABI {
```

Comment:
Similar comment here, is there any good reason not to already merge with the existing one used by lowering prepare?

Reply:
> you could annotations (TODOs/FIXMEs) for later fixing the parity issues.

Sounds good. I'll look into it.